subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbiExtractingClasspathResourceHasher.java

```diff
@@ -72,11 +71,11 @@ public HashCode hash(RegularFileSnapshot fileSnapshot) {
     }
 
     @Override
-    public HashCode hash(ZipEntry zipEntry, InputStream zipInput) throws IOException {
+    public HashCode hash(ZipEntry zipEntry) throws IOException {
         if (!isClassFile(zipEntry.getName())) {
             return null;
         }
-        return hashClassBytes(zipInput);
+        return hashClassBytes(zipEntry.getInputStream());
```

Comment:
We could optimize `hashClassBytes` by introducing the zip file size as an additional parameter, like:
```java
@Nullable
private HashCode hashClassBytes(InputStream inputStream, int size) throws IOException {
    // Use the ABI as the hash
    ClassReader reader = getClassReader(inputStream, size);
    ApiClassExtractor extractor = new ApiClassExtractor(Collections.<String>emptySet());
    if (extractor.shouldExtractApiClassFrom(reader)) {
        byte[] signature = extractor.extractApiClassFrom(reader);
        if (signature != null) {
            return Hashing.hashBytes(signature);
        }
    }
    return null;
}

private static ClassReader getClassReader(InputStream inputStream, int size) throws IOException {
    byte[] classBytes = new byte[size];
    ByteStreams.readFully(inputStream, classBytes);
    return new ClassReader(classBytes);
}
```
My measurements show a speed increase from your 16 seconds to 14 seconds this way.